### ISSUE TITLE
[REVIEW]: Catwalk.jl: An adaptive dispatch optimizer

### ISSUE THREAD

**Issue by whedon** [JuliaCon/proceedings-review#98]
**Submitting author:** @tisztamo (<a href="http://orcid.org/0000-0003-4931-4791">Krisztián Schäffer</a>)
**Repository:** <a href="https://github.com/tisztamo/Catwalk.jl" target ="_blank">https://github.com/tisztamo/Catwalk.jl</a>
**Version:** paper_juliacon_21
**Reviewer:** @aviatesk, @roger-luo
**Archive:** <a href="https://doi.org/10.5281/zenodo.6300994" target="_blank">10.5281/zenodo.6300994</a>

## Status

[![status](https://proceedings.juliacon.org/papers/4c71770c02886a29c0873745e7b84733/status.svg)](https://proceedings.juliacon.org/papers/4c71770c02886a29c0873745e7b84733)

Status badge code:

```
HTML: <a href="https://proceedings.juliacon.org/papers/4c71770c02886a29c0873745e7b84733"><img src="https://proceedings.juliacon.org/papers/4c71770c02886a29c0873745e7b84733/status.svg"></a>
Markdown: [![status](https://proceedings.juliacon.org/papers/4c71770c02886a29c0873745e7b84733/status.svg)](https://proceedings.juliacon.org/papers/4c71770c02886a29c0873745e7b84733)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/tisztamo/Catwalk.jl" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@aviatesk & @roger-luo, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @ranjanan know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JuliaCon conflict of interest policy](https://proceedings.juliacon.org/guide/authors#coi) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JuliaCon code of conduct](https://proceedings.juliacon.org/coc).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/tisztamo/Catwalk.jl">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Authorship:** Has the submitting author (@tisztamo) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Paper format

- [x] **Authors:** Does the `paper.tex` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?
- [x] **Page limit:** Is the page limit for extended abstracts respected by the submitted document?

### Content

- [x] **Context:** is the scientific context motivating the work correctly presented?
- [x] **Methodology:** is the approach taken in the work justified, presented with enough details and reference to reproduce it?
- [x] **Results:** are the results presented and compared to approaches with similar goals?
### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JuliaCon conflict of interest policy](https://proceedings.juliacon.org/guide/authors#coi) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JuliaCon code of conduct](https://proceedings.juliacon.org/coc).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/tisztamo/Catwalk.jl">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Authorship:** Has the submitting author (@tisztamo) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Paper format

- [x] **Authors:** Does the `paper.tex` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?
- [x] **Page limit:** Is the page limit for extended abstracts respected by the submitted document?

### Content

- [x] **Context:** is the scientific context motivating the work correctly presented?
- [x] **Methodology:** is the approach taken in the work justified, presented with enough details and reference to reproduce it?
- [x] **Results:** are the results presented and compared to approaches with similar goals?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @aviatesk, @roger-luo it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/JuliaCon/proceedings-review) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/JuliaCon/proceedings-review:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.08 s (452.5 files/s, 26204.7 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Julia                           20            218            306            877
Markdown                         5             81              0            260
TOML                             5             42              1            138
YAML                             5              3             12            129
CSS                              1              3              0             15
-------------------------------------------------------------------------------
SUM:                            36            347            319           1419
-------------------------------------------------------------------------------


Statistical information for the repository '7fc1f831a495b41f01f3221c' was
gathered on 2021/12/02.
No commited files with the specified extensions were found.
```

**Comment by ranjanan** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/JuliaCon/proceedings-papers/jcon.00098/jcon.00098/10.21105.jcon.00098.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/JuliaCon/proceedings-papers/blob/jcon.00098/jcon.00098/10.21105.jcon.00098.pdf) :page_facing_up: :point_left:

**Comment by ranjanan** [comment]
@roger-luo and @aviatesk a checklist has been added for each of you. Feel free to leave your reviewer comments here.

**Comment by Roger-luo** [comment]
This package looks nice in general to me (I actually tried it when it was first announced in discourse), the interface is simple to use and the performance gain is fantastic. The only thing that is missing is the community guidelines required above, I'd suggest just putting a https://github.com/SciML/ColPrac badge in the README

**Comment by tisztamo** [comment]
Thank you very much @Roger-luo for the review!

fix of the missing guideline: https://github.com/tisztamo/Catwalk.jl/commit/300518aaa0fed31d7a5f25a2fb5de220eec8a11c

**Comment by aviatesk** [comment]
I read the article, and here are few feedbacks on details:
- For the sentence "Types injected from dependent packages cannot benefit from the speedup" in `1. Introduction`: Union-splitting is applied for general Julia programs, so it optimizes arbitrary code. So I'd say the reference of `[1]` isn't much appropriate there
  - also, I couldn't find a reachable reference of `[2] Singledispatcharrays.jl`/`[5] Singledispatcharrays.jl` anymore
- maybe it's more appropriated if `2. Integration: Iterated staging` has some more detailed explanation of the usage (like as explained at <https://tisztamo.github.io/Catwalk.jl/dev/usage/>)?

---

As a general feedback, it might be more interesting if we can explain the essential difference between union-splitting (static optimization of runtime dispatches) vs. Catwalk's profiling based optimization.
The biggest difference would be that the former can only optimize it when all possible types are known AOT, while the latter can interact with runtime information. This means, in some situations, union-splitting is just enough. For example, if we change the definition of `xs` in [Catwalk's documentation](https://tisztamo.github.io/Catwalk.jl/dev/usage/) to `const xs = Union{Int,Float64,ComplexF64}[1, 2.0, ComplexF64(3.0, 3.0)]`, there is no essential performance difference:
```julia
julia> const NUM_BATCHES = 1000
1000

julia> function runbatches()
           for batchidx = 1:NUM_BATCHES
               hotloop()
               # Log progress, etc.
           end
       end
runbatches (generic function with 1 method)

julia> 

julia> const NUM_ITERS_PER_BATCH = 1_000_000
1000000

julia> function hotloop()
           for i = 1:NUM_ITERS_PER_BATCH
               x = get_some_x(i)
               calc_with_x(x)
           end
       end
hotloop (generic function with 1 method)

julia> const xs = Union{Int,Float64,ComplexF64}[1, 2.0, ComplexF64(3.0, 3.0)]
3-element Vector{Union{Float64, Int64, ComplexF64}}:
     1
     2.0
 3.0 + 3.0im

julia> get_some_x(i) = xs[i % length(xs) + 1]
get_some_x (generic function with 1 method)

julia> const result = Ref(ComplexF64(0.0, 0.0))
Base.RefValue{ComplexF64}(0.0 + 0.0im)

julia> function calc_with_x(x)
           result[] += x
       end
calc_with_x (generic function with 1 method)

julia> using Catwalk

julia> @jit calc_with_x x function hotloop_jit(jitctx)
           for i = 1:NUM_ITERS_PER_BATCH
               x = get_some_x(i)
               calc_with_x(x)
           end
       end
hotloop_jit (generic function with 1 method)

julia> function runbatches_jit()
           jit = Catwalk.JIT() ## Also works inside a function (no eval used)
           for batch = 1:NUM_BATCHES
               Catwalk.step!(jit)
               hotloop_jit(Catwalk.ctx(jit)) # Dynamic dispatch is "pulled out" from the hot loop to this call
           end
       end
runbatches_jit (generic function with 1 method)

julia> result[] = ComplexF64(0, 0)
0.0 + 0.0im

julia> @time runbatches_jit()
  4.193906 seconds (2.63 M allocations: 145.252 MiB, 0.50% gc time, 25.96% compilation time)

julia> jit_result = result[]
2.0e9 + 9.99999e8im

julia> result[] = ComplexF64(0, 0)
0.0 + 0.0im

julia> @time runbatches()
  2.854675 seconds
```
And then we may also want to mention the heuristic of Julia compiler, e.g. we should use Catwalk when we have more than [5 possible types](https://github.com/JuliaLang/julia/blob/83d7ebaca528904e7bcfc7f032f19d44fc1acd49/base/compiler/types.jl#L122) :)

**Comment by aviatesk** [comment]
If you like a more complicated story, we can also mention a part of Julia compiler where it does some sort of statical profiling JIT. It actually sorts method table by the runtime counts of dispatch, so Julia runtime actually does something very similar to Catwalk's idea (well, even in that case `isa`-based code generated by Catwalk is far more efficient than the whole method dispatch). So we can also make an advanced discussion on the computation costs of `isa` checks and runtime dispatch invocation, where the latter is around 20 CPU-cycles when dealing with a single argument, while the former is around 1-2 CPU cycles when consulted with a concrete type, if I remember correctly.

**Comment by tisztamo** [comment]
Thank you very much for your valuable comments, @aviatesk !

> * For the sentence "Types injected from dependent packages cannot benefit from the speedup" in `1. Introduction`: Union-splitting is applied for general Julia programs, so it optimizes arbitrary code. So I'd say the reference of `[1]` isn't much appropriate there

Right, this is not true for union splitting, I have to correct it (true for ManualDispatch.jl).

Still the inability to extend unions is prohibitive in some cases. Your later wording "can only optimize it when all possible types are known AOT" better describes the difference, but the precise meaning of AOT is not exactly clear to me. If we assume that a package may implement iterated staging combined with union split-based code generation - which it can do in a few dozen lines, although no package did it before Catwalk - , then the only differences are IIUC the limit of 5, and that Union{A, B, Any} will not be split. Do you mean this latter limitation with "AOT"?

> * also, I couldn't find a reachable reference of `[2] Singledispatcharrays.jl`/`[5] Singledispatcharrays.jl` anymore

It seems that the url field of references is not rendered, I will fix it. 

https://github.com/tkoolen/TypeSortedCollections.jl
https://github.com/melonedo/SingleDispatchArrays.jl

> in some situations, union-splitting is just enough

Yes, that is a contrived example, I will add a note to the docs. (The [Alternatives](https://tisztamo.github.io/Catwalk.jl/dev/#Alternatives) section of the introduction mentions union splitting among other solutions.)

> And then we may also want to mention the heuristic of Julia compiler, e.g. we should use Catwalk when we have more than [5 possible types](https://github.com/JuliaLang/julia/blob/83d7ebaca528904e7bcfc7f032f19d44fc1acd49/base/compiler/types.jl#L122) :)

I will find space for a sentence about it. (The one-page limitation makes it very hard to speak about other solutions, I thought that referencing them is enough.)

> Julia runtime actually does something very similar to Catwalk's idea

I did not know of the method table sorting, thanks for mentioning it! As I am not allowed to run out of the page - or I need to write 4+ more which I cannot do now -, unfortunately I have to leave it out for now.

> So we can also make an advanced discussion on the computation costs of isa checks and runtime dispatch invocation

The default cost model of Catwalk ( https://tisztamo.github.io/Catwalk.jl/dev/tuning/#Customize-the-cost-model ) is ad-hoc, it would be great to improve it! Discussing the relevant parts of the Julia compiler in a paper seems also helpful for the community as there is hard to find information on the internals. I hope to find time for a next iteration of Catwalk during the next year and I will contact you before starting if you don't mind.
 

Do you find the here outlined changes satisfactory?

Thank you for your help!

**Comment by aviatesk** [comment]
Yep, I think what you outlined above sounds good as a whole. I actually didn't think about the page limitation at all, so please summarize as you like -- my feedbacks are just suggestions :) 

> Do you mean this latter limitation with "AOT"?

By "AOT", I simply meant when Julia compiler compiles a hot function in question.
There are few cases when it fails to union-split (i.e. optimize abstract callsite):
- when the type in question is an "abstract" type, e.g. `Any`, `Number`, `Union{Number,String}`
- when the type is union and it needs to be more than 5 concrete types when split

In these cases, I think Catwalk can optimize them further by iterated staging using runtime information.

> Discussing the relevant parts of the Julia compiler in a paper seems also helpful for the community as there is hard to find information on the internals. I hope to find time for a next iteration of Catwalk during the next year and I will contact you before starting if you don't mind.

Yep, that would be great. Looking forward to the collaboration :)

**Comment by tisztamo** [comment]
Its my turn, sorry for the delay. I will do the required changes this weekend.

**Comment by tisztamo** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/JuliaCon/proceedings-papers/jcon.00098/jcon.00098/10.21105.jcon.00098.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/JuliaCon/proceedings-papers/blob/jcon.00098/jcon.00098/10.21105.jcon.00098.pdf) :page_facing_up: :point_left:

**Comment by tisztamo** [comment]
I made the required changes to the text:

https://github.com/tisztamo/Catwalk.jl/commit/106b91d22f42a598355055a0a159ab4e92b5da1f#diff-983b58b8186b0a4ed7f280f258cdab3eb0dd7d5136f8ac361ba982a43cfb7136

and updated the refs so that urls are rendered:

https://github.com/tisztamo/Catwalk.jl/commit/599e7b232e93c929aa6d1a1b4bf46d87bf8caffd#diff-382635c9325bf3273d195ff1b8a44e5b11afd7d97addeb8863ea35feb98c1a07

There is also a minimal change to the keywords:

https://github.com/tisztamo/Catwalk.jl/commit/e93afe2ef95143225f2cb7bf2b873f8b10b97071#diff-382635c9325bf3273d195ff1b8a44e5b11afd7d97addeb8863ea35feb98c1a07

@aviatesk Sorry for the delay again and thank you for reviewing the changes!

**Comment by ranjanan** [comment]
@aviatesk are the changes satisfactory to you?

**Comment by aviatesk** [comment]
sorry for the delay, lgtm

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/JuliaCon/proceedings-papers/jcon.00098/jcon.00098/10.21105.jcon.00098.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/JuliaCon/proceedings-papers/blob/jcon.00098/jcon.00098/10.21105.jcon.00098.pdf) :page_facing_up: :point_left:

**Comment by ranjanan** [comment]
@whedon accept

**Comment by whedon** [comment]
To recommend a paper to be accepted use `@whedon recommend-accept`

**Comment by ranjanan** [comment]
@whedon recommend-accept

**Comment by whedon** [comment]
No archive DOI set. Exiting...

**Comment by ranjanan** [comment]
@whedon help

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by ranjanan** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List all of Whedon's capabilities
@whedon commands

# Assign a GitHub user as the sole reviewer of this submission
@whedon assign @username as reviewer

# Add a GitHub user to the reviewers of this submission
@whedon add @username as reviewer

# Re-invite a reviewer (if they can't update checklists)
@whedon re-invite @username as reviewer

# Remove a GitHub user from the reviewers of this submission
@whedon remove @username as reviewer

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

# Change editorial assignment
@whedon assign @username as editor

# Set the software archive DOI at the top of the issue e.g.
@whedon set 10.0000/zenodo.00000 as archive

# Set the software version at the top of the issue e.g.
@whedon set v1.0.1 as version

# Open the review issue
@whedon start review

EDITORIAL TASKS

# All commands can be run on a non-default branch, to do this pass a custom 
# branch name by following the command with `from branch custom-branch-name`.
# For example:

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Remind an author or reviewer to return to a review after a
# certain period of time (supported units days and weeks)
@whedon remind @reviewer in 2 weeks

# Ask Whedon to do a dry run of accepting the paper and depositing with Crossref
@whedon recommend-accept

# Ask Whedon to check the references for missing DOIs
@whedon check references

# Ask Whedon to check repository statistics for the submitted software
@whedon check repository

EiC TASKS

# Invite an editor to edit a submission (sending them an email)
@whedon invite @editor as editor

# Reject a paper
@whedon reject

# Withdraw a paper
@whedon withdraw

# Ask Whedon to actually accept the paper and deposit with Crossref
@whedon accept deposit=true

```

**Comment by ranjanan** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Attempting to check references... from custom branch paper
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- None

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by tisztamo** [comment]
https://zenodo.org/record/6300994

Thank you very much!

**Comment by ranjanan** [comment]
@whedon set 10.5281/zenodo.6300994 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.6300994" target="_blank">10.5281/zenodo.6300994</a> is the archive.

**Comment by ranjanan** [comment]
@whedon set paper_juliacon_21 as the version

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by ranjanan** [comment]
@whedon set paper_juliacon_21 as version

**Comment by whedon** [comment]
OK. paper_juliacon_21 is the version.

**Comment by ranjanan** [comment]
@whedon recommend-accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #98 with the following error: 
```
 Can't find any papers to compile :-(

```

**Comment by ranjanan** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/JuliaCon/proceedings-papers/jcon.00098/jcon.00098/10.21105.jcon.00098.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/JuliaCon/proceedings-papers/blob/jcon.00098/jcon.00098/10.21105.jcon.00098.pdf) :page_facing_up: :point_left:

**Comment by ranjanan** [comment]
@whedon recommend-accept from branch paper

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- None

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @JuliaCon/jcon-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/JuliaCon/proceedings-papers/pull/55

If the paper PDF and Crossref deposit XML look good in https://github.com/JuliaCon/proceedings-papers/pull/55, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true from branch paper 
```

**Comment by ranjanan** [comment]
@tisztamo it looks on the proof that your header is dated 2019 (check page 2). Could you please update it to 21? I'll then accept this.

**Comment by tisztamo** [comment]
@ranjanan Could you please help me where to edit?

```
paper % grep 2019 *
journal_dat.tex:\def\@year{2019}
juliacon.cls:\def\filedate{2019 04 07}
Binary file logojuliacon.pdf matches
paper.log:Package: bm 2019/07/24 v1.2d Bold Symbol Support (DPC/FMi)
...
paper.log:Package: gettitlestring 2019/12/15 v1.6 Cleanup title references (HO)
prep.rb:year = ENV["JLCON_YEAR"] === nil ? 2019 : ENV["JLCON_YEAR"]
```

`journal_dat.tex` is generated. Should I edit the default in prep.rb? (It says "DO NOT EDIT").

**Comment by ranjanan** [comment]
You could just recompile with the new template which is dated 2021. That may be easiest.

**Comment by ranjanan** [comment]
@tisztamo bump

**Comment by tisztamo** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/JuliaCon/proceedings-papers/jcon.00098/jcon.00098/10.21105.jcon.00098.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/JuliaCon/proceedings-papers/blob/jcon.00098/jcon.00098/10.21105.jcon.00098.pdf) :page_facing_up: :point_left:

**Comment by tisztamo** [comment]
@ranjanan The year is 2021 now. Sorry for being late again!

**Comment by ranjanan** [comment]
@whedon generate pdf from branch paper

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/JuliaCon/proceedings-papers/jcon.00098/jcon.00098/10.21105.jcon.00098.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/JuliaCon/proceedings-papers/blob/jcon.00098/jcon.00098/10.21105.jcon.00098.pdf) :page_facing_up: :point_left:

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
:wave: @JuliaCon/jcon-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/JuliaCon/proceedings-papers/pull/56

If the paper PDF and Crossref deposit XML look good in https://github.com/JuliaCon/proceedings-papers/pull/56, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true from branch paper 
```

**Comment by whedon** [comment]
I'm sorry @ranjanan, I'm afraid I can't do that. That's something only editor-in-chiefs are allowed to do.

**Comment by carstenbauer** [comment]
@whedon accept deposit=true from branch paper

**Comment by whedon** [comment]
I'm sorry @carstenbauer, I'm afraid I can't do that. That's something only editor-in-chiefs are allowed to do.

**Comment by vchuravy** [comment]
@whedon accept deposit=true from branch paper

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JCON!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/JuliaCon/proceedings-papers/pull/57
1. Wait a couple of minutes, then verify that the paper DOI resolves [https://doi.org/10.21105/jcon.00098](https://doi.org/10.21105/jcon.00098)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://proceedings.juliacon.org/papers/10.21105/jcon.00098/status.svg)](https://doi.org/10.21105/jcon.00098)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/jcon.00098">
  <img src="https://proceedings.juliacon.org/papers/10.21105/jcon.00098/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://proceedings.juliacon.org/papers/10.21105/jcon.00098/status.svg
   :target: https://doi.org/10.21105/jcon.00098
```
This is how it will look in your documentation:

[![DOI](https://proceedings.juliacon.org/papers/10.21105/jcon.00098/status.svg)](https://doi.org/10.21105/jcon.00098)

**We need your help!**

JuliaCon Proceedings is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://forms.gle/yANMe4c9YmuW9MWcA
- Making a small donation to support our running costs here: https://numfocus.salsalabs.org/donate-to-jcon

**Comment by ranjanan** [comment]
Thanks @tisztamo!
